rfcs/0012-declarative-virtual-machines.md

```diff
@@ -0,0 +1,249 @@
+want the same ease of use as with containers.
+
+## Expected outcome
+
+This RFC describes a prospective NixOS module for running VMs. Thus, the main
+design point is that a `vms.` module should be made available to NixOS users.
+It should allow to declaratively define virtual machines, including their
+configuration.
+
+# Detailed design
+[design]: #detailed-design
+
+The VM module is architectured in the following way:
+ * The root filesystem is written on a qcow2 image
+ * The store is stored on the host, for easy upgrade handling (see [Disk
+   management](#disk-management))
+ * Each VM is run as a (set of) systemd service, and can be rebooted using
+   `systemctl restart vm-${name}.service`
+ * `qemu` fetches the kernel and initrd directly from the guests' store
+ * Networking is done with all the VMs in a bridge internal to the server (see
+   [Networking](#networking))
+
+Details follow.
+
+## Basic configuration
+
+The user should be able to configure various aspects of the VMs running,
+especially disk, memory and vcpu usages.
+
+### Proposed option set
+
+The following options are proposed:
+
+```nix
+{
+  vms = {
+    path = "/path/to/dir"; # Path into which to store persistent data (disk
+                           # images and per-vm store)
+    rpath = "/runtime/dir"; # Path for temporary non-user-facing low-size data,
+                            # like IPC sockets
+
+    machines.${name} = {
+      diskSize = 10240; # Size (in MiB) of the disk image excluding shared paths
+                        # and store
+      memorySize = 1024; # Size (in MiB) of the RAM allocated to the VM
+      vcpus = 1; # Number of virtual CPUs the VM will see
+    };
+  };
+}
+```
+
+## Disk management
+[disk-management]: #disk-management
+
+The VM must have its own disk image, yet must also have shared access to folders
+on the host, if the configuration dictates it so.
+
+The store is a special case, as it has to be written by the host, in order to
+handle both initial VM creation and upgrades, and read by the guest. As a
+consequence, the easy way of doing it is to bind it as a virtfs between the host
+and the guest. The host writes new packages on it, and the guest is able to
+execute them.
+
+The aim is to be able to trigger an update without rebooting the guest, by
+having the host write the additional packages to the guest's nix store and call
+the `switch-to-configuration` script on the guest in some way. However, this is
+not included in the current RFC, for the sake of simplicity: for now, upgrading
+is done by pointing qemu at the new configuration (by writing the new systemd
+service), and rebooting the VM (by restarting the systemd service).
+
+In order to do this, a possible way to do so is to mount:
+ * `/` as a filesystem on a qcow2 image
+ * `/nix/store` as a virtfs onto a directory on the host, in order to easily
+   handle setup and upgrades from the host
```

Comment:
I didn't mention explicitly that the directory on the host it's mapped to is not the host's /nix/store, fixed now, thanks!